File: tests/game_objects/test_components.py
"""Tests all classes and functions in game_objects/components.py."""
from __future__ import annotations

# Pip
import pytest

# Custom
from hades.game_objects.components import GameObjectAttributeBase

__all__ = ()


class TestGameObjectAttribute(GameObjectAttributeBase):
    """Represents a game object attribute useful for testing."""


@pytest.fixture()
def game_object_attribute() -> TestGameObjectAttribute:
    """Create a game object attribute for use in testing.

    Returns:
        The game object attribute for use in testing.
    """
    return TestGameObjectAttribute(150, 3)


def test_game_object_attribute_setter_higher(
    game_object_attribute: TestGameObjectAttribute,
) -> None:
    """Test that a game object attribute is set with a higher value correctly.

    Args:
        game_object_attribute: The game object attribute for use in testing.
    """
    game_object_attribute.value = 200
    assert game_object_attribute.value == 150


def test_game_object_attribute_setter_lower(
    game_object_attribute: TestGameObjectAttribute,
) -> None:
    """Test that a game object attribute is set with a lower value correctly.

    Args:
        game_object_attribute: The game object attribute for use in testing.
    """
    game_object_attribute.value = 100
    assert game_object_attribute.value == 100


def test_game_object_attribute_setter_iadd(
    game_object_attribute: TestGameObjectAttribute,
) -> None:
    """Test that adding a value to the game object attribute is correct.

    Args:
        game_object_attribute: The game object attribute for use in testing.
    """
    game_object_attribute.value += 100
    assert game_object_attribute.value == 150


def test_game_object_attribute_setter_isub(
    game_object_attribute: TestGameObjectAttribute,
) -> None:
    """Test that subtracting a value from the game object attribute is correct.

    Args:
        game_object_attribute: The game object attribute for use in testing.
    """
    game_object_attribute.value -= 200
    assert game_object_attribute.value == 0

File: hades_extensions/test/game_objects/test_components.cpp
// External includes
#include "gtest/gtest.h"

// Custom includes
#include "game_objects/components.hpp"

// ----- CLASSES ------------------------------
/// Represents a game object attribute useful for testing.
class TestGameObjectAttribute : public GameObjectAttributeBase {
 public:
  /// Initialise the object.
  ///
  /// @param initial_value - The initial value of the movement force attribute.
  /// @param level_limit - The level limit of the movement force attribute.
  TestGameObjectAttribute(float initial_value, int level_limit) : GameObjectAttributeBase(initial_value, level_limit) {}
};

// ----- FIXTURES ------------------------------
/// A test fixture for the game_objects/components.hpp tests.
class ComponentsFixtures : public testing::Test {
 protected:
  /// A test game object attribute.
  TestGameObjectAttribute test_game_object_attribute{150, 3};
};

// ----- TESTS ------------------------------
/// Test that a game object attribute is set with a higher value correctly.
TEST_F(ComponentsFixtures, TestGameObjectAttributeSetterHigher) {
  test_game_object_attribute.value(200);
  ASSERT_EQ(test_game_object_attribute.value(), 150);
}

/// Test that a game object attribute is set with a lower value correctly.
TEST_F(ComponentsFixtures, TestGameObjectAttributeSetterLower) {
  test_game_object_attribute.value(100);
  ASSERT_EQ(test_game_object_attribute.value(), 100);
}

/// Test that adding a value to the game object attribute is correct.
TEST_F(ComponentsFixtures, TestGameObjectAttributeSetterAdd) {
  test_game_object_attribute.value(test_game_object_attribute.value() + 100);
  ASSERT_EQ(test_game_object_attribute.value(), 150);
}

/// Test that subtracting a value from the game object attribute is correct.
TEST_F(ComponentsFixtures, TestGameObjectAttributeSetterSub) {
  test_game_object_attribute.value(test_game_object_attribute.value() - 200);
  ASSERT_EQ(test_game_object_attribute.value(), 0);
}

/// Test that the inventory capacity calculation is correct.
TEST(Tests, TestInventoryCapacity) {
  // Test if an inventory with zero width and height has a zero capacity
  Inventory inventory_zero_capacity{0, 0};
  ASSERT_EQ(inventory_zero_capacity.capacity(), 0);

  // Test if an inventory with zero width and non-zero height has a zero
  // capacity
  Inventory inventory_zero_width{0, 10};
  ASSERT_EQ(inventory_zero_width.capacity(), 0);

  // Test if an inventory with non-zero width and zero height has a zero
  // capacity
  Inventory inventory_zero_height{10, 0};
  ASSERT_EQ(inventory_zero_height.capacity(), 0);

  // Test if an inventory with non-zero width and non-zero height has a
  // non-zero capacity
  Inventory inventory_non_zero{5, 10};
  ASSERT_EQ(inventory_non_zero.capacity(), 50);
}
